src/jit/lsraxarch.cpp

```diff
@@ -360,47 +438,85 @@ void LinearScan::BuildNode(GenTree* tree)
-            LocationInfoListNode* valueInfo = getLocationInfo(tree->gtCmpXchg.gtOpValue);
-            valueInfo->info.setSrcCandidates(this, allRegs(TYP_INT) & ~RBM_RAX);
-            useList.Append(valueInfo);
-            info->setDstCandidates(this, RBM_RAX);
-            LocationInfoListNode* comparandInfo = getLocationInfo(tree->gtCmpXchg.gtOpComparand);
-            comparandInfo->info.setSrcCandidates(this, RBM_RAX);
-            useList.Append(comparandInfo);
+            srcCount = 3;
+            assert(dstCount == 1);
+
+            // Comparand is preferenced to RAX.
+            // The remaining two operands can be in any reg other than RAX.
+            BuildUse(tree->gtCmpXchg.gtOpLocation, allRegs(TYP_INT) & ~RBM_RAX);
+            BuildUse(tree->gtCmpXchg.gtOpValue, allRegs(TYP_INT) & ~RBM_RAX);
+            BuildUse(tree->gtCmpXchg.gtOpComparand, RBM_RAX);
+            BuildDef(tree, RBM_RAX);
         }
         break;
 
+        case GT_XADD:
+        case GT_XCHG:
         case GT_LOCKADD:
-            info->srcCount = appendBinaryLocationInfoToList(tree->AsOp());
-            assert(info->dstCount == (tree->TypeGet() == TYP_VOID) ? 0 : 1);
-            break;
+        {
+            // TODO-XArch-Cleanup: We should make the indirection explicit on these nodes so that we don't have
+            // to special case them.
+            // These tree nodes will have their op1 marked as isDelayFree=true.
+            // That is, op1's reg remains in use until the subsequent instruction.
+            GenTree* addr = tree->gtOp.gtOp1;
+            GenTree* data = tree->gtOp.gtOp2;
+            assert(!addr->isContained());
+            RefPosition* addrUse = BuildUse(addr);
+            setDelayFree(addrUse);
+            tgtPrefUse = addrUse;
+            srcCount   = 1;
+            dstCount   = 1;
+            if (!data->isContained())
+            {
+                RefPosition* dataUse = dataUse = BuildUse(data);
+                srcCount                       = 2;
+            }
+
+            if (tree->TypeGet() == TYP_VOID)
+            {
+                // Right now a GT_XADD node could be morphed into a
+                // GT_LOCKADD of TYP_VOID. See gtExtractSideEffList().
+                // Note that it is advantageous to use GT_LOCKADD
+                // instead of of GT_XADD as the former uses lock.add,
+                // which allows its second operand to be a contained
+                // immediate wheres xadd instruction requires its
+                // second operand to be in a register.
+                // Give it an artificial type and mark it as an unused value.
+                // This results in a Def position created but not considered consumed by its parent node.
+                tree->gtType  = TYP_INT;
+                isLocalDefUse = true;
+                tree->SetUnusedValue();
+            }
+            RefPosition* def                       = BuildDef(tree);
```

Comment:
Vertical alignment of all assignments is the worst formatting rule the JIT code uses...